types/tablet_v2/wlr_tablet_v2_tool.c

```diff
@@ -0,0 +1,515 @@
+			ZWP_TABLET_TOOL_V2_CAPABILITY_SLIDER);
+	}
+
+	if (tool->wlr_tool->wheel) {
+		zwp_tablet_tool_v2_send_capability(client->resource,
+			ZWP_TABLET_TOOL_V2_CAPABILITY_WHEEL);
+	}
+
+	zwp_tablet_tool_v2_send_done(client->resource);
+
+	client->client = seat->wl_client;
+	wl_list_insert(&seat->tools, &client->seat_link);
+	wl_list_insert(&tool->clients, &client->tool_link);
+}
+
+static void handle_wlr_tablet_tool_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_tablet_v2_tablet_tool *tool =
+		wl_container_of(listener, tool, tool_destroy);
+
+	struct wlr_tablet_tool_client_v2 *pos;
+	struct wlr_tablet_tool_client_v2 *tmp;
+	wl_list_for_each_safe(pos, tmp, &tool->clients, tool_link) {
+		zwp_tablet_tool_v2_send_removed(pos->resource);
+		pos->tool = NULL;
+	}
+
+	wl_list_remove(&tool->clients);
+	wl_list_remove(&tool->link);
+	wl_list_remove(&tool->tool_destroy.link);
+	wl_list_remove(&tool->events.set_cursor.listener_list);
+	free(tool);
+}
+
+struct wlr_tablet_v2_tablet_tool *wlr_tablet_tool_create(
+		struct wlr_tablet_manager_v2 *manager,
+		struct wlr_seat *wlr_seat,
+		struct wlr_tablet_tool *wlr_tool) {
+	struct wlr_tablet_seat_v2 *seat = get_or_create_tablet_seat(manager, wlr_seat);
+	if (!seat) {
+		return NULL;
+	}
+	struct wlr_tablet_v2_tablet_tool *tool =
+		calloc(1, sizeof(struct wlr_tablet_v2_tablet_tool));
+	if (!tool) {
+		return NULL;
+	}
+
+	tool->wlr_tool = wlr_tool;
+	wl_list_init(&tool->clients);
+
+
+	tool->tool_destroy.notify = handle_wlr_tablet_tool_destroy;
+	wl_signal_add(&wlr_tool->events.destroy, &tool->tool_destroy);
+	wl_list_insert(&seat->tools, &tool->link);
+
+	// We need to create a tablet client for all clients on the seat
+	struct wlr_tablet_seat_client_v2 *pos;
+	wl_list_for_each(pos, &seat->clients, seat_link) {
+		// Tell the clients about the new tool
+		add_tablet_tool_client(pos, tool);
+	}
+
+	wl_signal_init(&tool->events.set_cursor);
+
+	return tool;
+}
+
+struct wlr_tablet_tool_client_v2 *tablet_tool_client_from_resource(struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwp_tablet_tool_v2_interface,
+		&tablet_tool_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+
+/* Actual protocol foo */
+// https://www.geeksforgeeks.org/move-zeroes-end-array/
+static size_t push_zeroes_to_end(uint32_t arr[], size_t n) {
+	size_t count = 0;
+
+	for (size_t i = 0; i < n; i++) {
+		if (arr[i] != 0) {
+			arr[count++] = arr[i];
+		}
+	}
+
+	size_t ret = count;
+
+	while (count < n) {
+		arr[count++] = 0;
+	}
+
+	return ret;
+}
+
+static ssize_t tablet_tool_button_update(struct wlr_tablet_v2_tablet_tool *tool,
+		uint32_t button, enum zwp_tablet_pad_v2_button_state state) {
+	bool found = false;
+	size_t i = 0;
+	for (; i < tool->num_buttons; ++i) {
+		if (tool->pressed_buttons[i] == button) {
+			found = true;
+			break;
+		}
+	}
+
+	if (state == ZWP_TABLET_PAD_V2_BUTTON_STATE_PRESSED && !found &&
+			tool->num_buttons < WLR_TABLET_V2_TOOL_BUTTONS_CAP) {
```

Comment:
We need to at least warn here if they try to press 17 buttons.

Reply:
Provided that we find a tablet tool like the Naga :)
But agreed, it should be warned about